### PR TITLE
ldc: 1.36.0 -> 1.38.0, clean up

### PR DESCRIPTION
**Pull request by jtbx**
## Description of changes

https://github.com/ldc-developers/ldc/releases/tag/v1.37.0

**(edit 2024/05/12)** https://github.com/ldc-developers/ldc/releases/tag/v1.38.0

Also migrate to pkgs/by-name and clean up the package a bit (see commits).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jtbx** [comment]
Waiting on #288841

**Comment by TomaSajt** [comment]
Just a heads up: the other PRs have been merged

**Comment by jtbx** [comment]
Thanks. 1.38.0 is almost ready, I'll wait for that first.

**Comment by jtbx** [comment]
Updated to 1.38.0

File: pkgs/by-name/ld/ldc/generic.nix [hunk 1]
```diff
@@ -14,52 +30,46 @@ let
...
-  ''
-  + lib.optionalString stdenv.hostPlatform.isLinux ''
-      substituteInPlace runtime/phobos/std/socket.d --replace "assert(ih.addrList[0] == 0x7F_00_00_01);" ""
-  ''
-  + lib.optionalString stdenv.hostPlatform.isDarwin ''
-      substituteInPlace runtime/phobos/std/socket.d --replace "foreach (name; names)" "names = []; foreach (name; names)"
+  postPatch = lib.optionalString stdenv.hostPlatform.isLinux ''
+    substituteInPlace runtime/phobos/std/socket.d --replace "assert(ih.addrList[0] == 0x7F_00_00_01);" ""
+  '' + lib.optionalString stdenv.hostPlatform.isDarwin ''
+    substituteInPlace runtime/phobos/std/socket.d --replace "foreach (name; names)" "names = []; foreach (name; names)"
```
Comment (TomaSajt) [inline pkgs/by-name/ld/ldc/generic.nix:65]:
```suggestion
  postPatch = lib.optionalString stdenv.hostPlatform.isLinux ''
    substituteInPlace runtime/phobos/std/socket.d --replace-fail "assert(ih.addrList[0] == 0x7F_00_00_01);" ""
  '' + lib.optionalString stdenv.hostPlatform.isDarwin ''
    substituteInPlace runtime/phobos/std/socket.d --replace-fail "foreach (name; names)" "names = []; foreach (name; names)"
```

**Review by TomaSajt — CHANGES_REQUESTED** [review]
Mostly LGTM, some changes requested

File: pkgs/by-name/ld/ldc/generic.nix (hunk 1 above)
Comment (jtbx) [reply pkgs/by-name/ld/ldc/generic.nix]: Thanks, applied

**Review by TomaSajt — APPROVED** [review]

**Comment by jtbx** [comment]
Result of `nixpkgs-review pr 293417` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>gtkd</li>
    <li>inochi-creator</li>
    <li>inochi-session</li>
    <li>tilix</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>Literate</li>
    <li>btdu</li>
    <li>cheesecutter</li>
    <li>dstep</li>
    <li>dtools</li>
    <li>dub</li>
    <li>ldc</li>
    <li>luneta</li>
    <li>onedrive</li>
    <li>onedrivegui</li>
    <li>onedrivegui.dist</li>
    <li>rund</li>
    <li>sambamba</li>
    <li>serve-d</li>
  </ul>
</details>

**Comment by jtbx** [comment]
Tilix fails to build because gtkd fails to build, and gtkd fails to build because
of something I'm guessing is in the fixup phase:
```
Running phase: fixupPhase
shrinking RPATHs of ELF executables and libraries in /nix/store/h69b9p2x7kfv6ijc5cxy8kkbifpdcgif-gtkd-3.10.0
checking for references to /tmp/nix-build-gtkd-3.10.0.drv-0/ in /nix/store/h69b9p2x7kfv6ijc5cxy8kkbifpdcgif-gtkd-3.10.0...
patching script interpreter paths in /nix/store/h69b9p2x7kfv6ijc5cxy8kkbifpdcgif-gtkd-3.10.0
stripping (with command strip and flags -S -p) in  /nix/store/h69b9p2x7kfv6ijc5cxy8kkbifpdcgif-gtkd-3.10.0/lib
chmod: changing permissions of '/lib/pkgconfig/atk-bridge-2.0.pc': Operation not permitted
/nix/store/558iw5j1bk7z6wrg8cp96q2rx03jqj1v-stdenv-linux/setup: line 131: pop_var_context: head of shell_variables not a function context
```

Inochi fails because the json library attempts to throw an immutable exception:
```
../.dub/packages/fghj/1.0.2/fghj/source/fghj/fghj.d(244,19): Deprecation: cannot throw object of qualified type `immutable(EmptyFghjException)`
```
[fghj.d(244)](https://github.com/Inochi2D/fghj/blob/main/source/fghj/fghj.d#L244)

I can't find anything in the changelogs for 2.107-2.108 about this either...

**Comment by TomaSajt** [comment]
If you look at the inochi logs, the errors are earlier.
And are results of this being fixed: https://issues.dlang.org/show_bug.cgi?id=24387 in `dmd`, which is in the `ldc` release

**Comment by jtbx** [comment]
Ah I didn't see that error, that makes a lot more sense now.

**Comment by TomaSajt** [comment]
Okay, so I came up with a fix for `inochi` stuff

This should go in the `postConfigure` of `pkgs/applications/misc/inochi2d/generic.nix`
```
inochi2d_dir=("$DUB_HOME"/packages/inochi2d/*/inochi2d)
patch -p1 -d "$inochi2d_dir" < ${./ldc-1.38.0-compat.patch}
```

Where `ldc-1.38.0-compat.patch` is the following:
```diff
diff --git a/source/inochi2d/core/nodes/drivers/package.d b/source/inochi2d/core/nodes/drivers/package.d
index 9a46846..01046b6 100644
--- a/source/inochi2d/core/nodes/drivers/package.d
+++ b/source/inochi2d/core/nodes/drivers/package.d
@@ -17,10 +17,8 @@ public import inochi2d.core.nodes.drivers.simplephysics;
 */
 @TypeId("Driver")
 abstract class Driver : Node {
-private:
-    this() { }
-
 protected:
+    this() { }
     /**
         Constructs a new Driver node
     */
diff --git a/source/inochi2d/core/nodes/package.d b/source/inochi2d/core/nodes/package.d
index 6090355..5045361 100644
--- a/source/inochi2d/core/nodes/package.d
+++ b/source/inochi2d/core/nodes/package.d
@@ -74,8 +74,6 @@ void inClearUUIDs() {
 @TypeId("Node")
 class Node : ISerializable {
 private:
-    this() { }
-
     @Ignore
     Puppet puppet_;
 
@@ -98,6 +96,8 @@ private:
     string nodePath_;
 
 protected:
+    this() { }
+
     bool preProcessed  = false;
     bool postProcessed = false;
 
```

**Comment by jtbx** [comment]
Awesome thanks! I put it in a separate commit with you set as the author if that's alright.

**Comment by jtbx** [comment]
I've found that the gtkd build failure is caused by the call `substituteInPlace` in the [postFixup hook](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/gtkd/default.nix#L117). `substituteInPlace` invokes `chmod` [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L914) and that's where it fails. Weird because this hasn't happened before...

I've tried reverting 9453650, which causes a new "cyclic reference" error.

**Comment by TomaSajt** [comment]
> I've found that the gtkd build failure is caused by the `substituteInPlace` call in the [postFixup hook](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/gtkd/default.nix#L117). `substituteInPlace` invokes `chmod` [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L914) and that's where it fails. Weird because this hasn't happened before...
> 
> I've tried reverting 9453650, which causes a new "cyclic reference" error.

What's weird is that I didn't have any issues compiling it. I'll try on my laptop as well, to see if that also works.

**Comment by jtbx** [comment]
I wouldn't be surprised if it's only failing on my system, this setup has a bit pretty problematic in the past. So much for reproduciblity :\

**Comment by jtbx** [comment]
Yeah the build succeeds on my NixOS box, I think this is good to go now.

**Comment by TomaSajt** [comment]
By the way, you should probably remove the draft status.

**Comment by jtbx** [comment]
Ah that's right. Thanks for all your help <3

File: pkgs/by-name/ld/ldc/generic.nix [hunk 2]
```diff
@@ -14,52 +30,46 @@ let
...
+    repo = "ldc";
+    rev = "v${version}";
+    inherit hash;
+    fetchSubmodules = true;
   };
 
   # https://issues.dlang.org/show_bug.cgi?id=19553
   hardeningDisable = [ "fortify" ];
 
   postUnpack = ''
```
Comment (TomaSajt) [inline pkgs/by-name/ld/ldc/generic.nix:48]: I think that you should move this to postPatch or something else that happens after the builder moves into `sourceRoot`.
This way you don't have rely on the fact that `fetchFromGitHub` has `source` as its derivation name.

**Review by TomaSajt — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ld/ldc/generic.nix (hunk 2 above)
Comment (jtbx) [reply pkgs/by-name/ld/ldc/generic.nix]: Done, thanks

**Comment by TomaSajt** [comment]
By the way, what was/is the reason behind extracting the `version` and `hash` into a different file?
If it was a multi-versioned package it would be understandable, however it is not.

**Comment by jtbx** [comment]
Good point, the dmd package does the same. I can change it back.

**Comment by jtbx** [comment]
@lionello is there any reason for the package versions to be in separate files?

**Comment by TomaSajt** [comment]
Now that https://github.com/NixOS/nixpkgs/pull/314953 is merged, the inochi patches can probably be dropped

**Comment by jtbx** [comment]
Yep, everything builds now:
```
18 packages built:
Literate btdu cheesecutter dstep dtools dub gtkd inochi-creator inochi-session ldc luneta onedrive onedrivegui onedrivegui.dist rund sambamba serve-d tilix
```

File: pkgs/by-name/ld/ldc/package.nix [hunk 3]
```diff
@@ -14,52 +29,44 @@ let
...
-    substituteInPlace tests/dmd/Makefile --replace "SHELL=/bin/bash" "SHELL=${bash}/bin/bash"
-  ''
-  + lib.optionalString stdenv.hostPlatform.isLinux ''
-      substituteInPlace runtime/phobos/std/socket.d --replace "assert(ih.addrList[0] == 0x7F_00_00_01);" ""
-  ''
-  + lib.optionalString stdenv.hostPlatform.isDarwin ''
-      substituteInPlace runtime/phobos/std/socket.d --replace "foreach (name; names)" "names = []; foreach (name; names)"
+    rm tests/dmd/fail_compilation/mixin_gc.d
+    rm tests/dmd/runnable/xtest46_gc.d
+    rm tests/dmd/runnable/testptrref_gc.d
```
[the remark this answers is not in the document]
Comment (jtbx) [reply pkgs/by-name/ld/ldc/package.nix:52]: After a lot of digging I found that these lines were introduced in 31fe00df8ac21c1594448a82cd95d7c69d014b60 without reason (nothing in #59197 either). Often there are some tests for LDC which fail when running under Nix.

File: pkgs/development/compilers/ldc/bootstrap.nix [hunk 4]
```diff
@@ -8,4 +12,39 @@ callPackage ./binary.nix {
...
     osx-arm64  = "sha256-Nb/owBdIeroB9jLMDvwjo8bvsTC9vFyJPLMTOMsSAd4=";
   };
+in stdenv.mkDerivation {
+  pname = "ldc-bootstrap";
+  inherit version;
+
+  src = fetchurl rec {
+    name = "ldc2-${version}-${OS}-${ARCH}.tar.xz";
+    url = "https://github.com/ldc-developers/ldc/releases/download/v${version}/${name}";
+    sha256 = hashes."${OS}-${ARCH}" or (throw "missing bootstrap sha256 for ${OS}-${ARCH}");
```
[the remark this answers is not in the document]
Comment (jtbx) [reply pkgs/development/compilers/ldc/bootstrap.nix]: Applied.

File: pkgs/by-name/ld/ldc/package.nix [hunk 5]
```diff
@@ -14,52 +29,44 @@ let
...
+    hash = "sha256-d/UREh+fHRRh0r3H60uPjHute+qspcm9TBFWZMbGDxk=";
+    fetchSubmodules = true;
   };
 
   # https://issues.dlang.org/show_bug.cgi?id=19553
   hardeningDisable = [ "fortify" ];
 
-  postUnpack = ''
+  postPatch = ''
     patchShebangs .
```
[the remark this answers is not in the document]
Comment (jtbx) [reply pkgs/by-name/ld/ldc/package.nix]: Only scripts in three different directories are being patched. Pushed!

**Comment by jtbx** [comment]
Result of `nixpkgs-review pr 293417` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>inochi-session</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>Literate</li>
    <li>btdu</li>
    <li>dstep</li>
    <li>dtools</li>
    <li>dub</li>
    <li>gtkd</li>
    <li>inochi-creator</li>
    <li>ldc</li>
    <li>luneta</li>
    <li>onedrive</li>
    <li>onedrivegui</li>
    <li>onedrivegui.dist</li>
    <li>rund</li>
    <li>serve-d</li>
    <li>tilix</li>
  </ul>
</details>

**Comment by TomaSajt** [comment]
> Result of `nixpkgs-review pr 293417` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
> <details>
>   <summary>1 package marked as broken and skipped:</summary>
>   <ul>
>     <li>inochi-session</li>
>   </ul>
> </details>
> <details>
>   <summary>15 packages built:</summary>
>   <ul>
>     <li>Literate</li>
>     <li>btdu</li>
>     <li>dstep</li>
>     <li>dtools</li>
>     <li>dub</li>
>     <li>gtkd</li>
>     <li>inochi-creator</li>
>     <li>ldc</li>
>     <li>luneta</li>
>     <li>onedrive</li>
>     <li>onedrivegui</li>
>     <li>onedrivegui.dist</li>
>     <li>rund</li>
>     <li>serve-d</li>
>     <li>tilix</li>
>   </ul>
> </details>

I read something about aarch64 support being dropped for inochi, but thought it would be only for next major release.
One of the two was already failing for aarch64 so it was already marked as broken.
Perhaps the supported platforms should be narrowed instead. I'll look into it later.

**Review by pbsds — APPROVED** [review]

**Comment by pbsds** [comment]
> I read something about aarch64 support being dropped for inochi

Asahi Lina would never let that happen ;)

**Comment by jtbx** [comment]
Thank you!

**Pull request merged**